### ISSUE TITLE
Listen to HLS.js errors internally

### ISSUE THREAD

**Issue by nnarhinen** [flowplayer/flowplayer#1331]
Currently we only listen to video tag errors and we miss on the fine grained error handling HLS.js offers.

**Comment by codener** [comment]
We have dug into error handling by the `hlsjs` engine. This is how we believe error handling should happen with the `hlsjs-lite` engine:

### How error handling works in Hls.js:

- hls.js can recover from "network errors" and "media errors". These are error types that are both encountered in Hls.js internally and the <video> element

- hls.js automatically recovers from both errors internally for a configured number of retries. For each retry a non-fatal error is triggered on the Hls JS object. The developer would usually ignore this error, since Hls.js is still retrying internally.

- When that number of retries is exceeded, an fatal error on the Hls object is triggered (`fatal` property true on error object). The developer can then choose what to do with a fatal error. An (incomplete) suggestion for a default implementation is here: https://github.com/video-dev/hls.js/blob/master/doc/API.md#hlsrecovermediaerror


Unfortunately the hlsjs-lite engine has lost much of the functionality of flowplayer-hlsjs.
Inparticular, error recovery code is missing completely.

### We think this is the minimum hlsjs-lite should do to implement basic error recovery: 

- When Hls.js signals a non-fatal network error:
  - Don't do anything, Hls.js will retry internally
  - Don't trigger Flowplayer error

- When Hls.js signals a fatal network error:
  - Call hls.startLoad();
  - Don't trigger Flowplayer error
  - Throttle recovery attempts to 3 seconds
  - We should give up on recovery after a configurable number of seconds and trigger a Flowplayer error
  
- When Hls.js signals a non-fatal media error:
  - Don't do anything, Hls.js will retry internally
  - Don't trigger Flowplayer error
  
- When Hls.js signals a fatal media error:
  - Call hls.recoverMediaError();
  - Don't trigger Flowplayer error
  - Throttle recovery attempts to 3 seconds
  - If recovery didn't work, call hls.swapAudioCodec() before hls.recoverMediaError() on the next attempt
  - We should give up on recovery after a configurable number of seconds and trigger a Flowplayer error
  
- If there is an error on the video element:
  - We need to listen to its `error` event
  - We need to check if this is an error that we will recover from (look at mediaElement.error.code)
  - If our recover code will run, we must preventDefault() on that error event
  

### Code
```
errorHandling = (hls, player, root) ->
  // Called after html5factory has created a new <video> element
  // Caution: "ext" is the engine once vs extensions in these links below, taking different arguments
  // https://github.com/flowplayer/flowplayer/blob/9ca2d2d616ca974f9dcd624bf96a27c77461597f/lib/engine/html5-factory.js#L113
  // https://github.com/flowplayer/flowplayer/blob/482a7fa1952f159c74fcb567b27087ba639b1b1b/lib/engine/hlsjs.js#L25
  
  // Unfortunately html5factory doesn't pass us the video element
  video = root.querySelector('video')
  
  hls.on Hls.Events.ERROR, (event) ->
    // "fancy" edge case handling here
    
  video.addEventListener 'error', (event) ->
    if event.error.code == CODE_DEN_WIR_HANDELN
      event.preventDefault()
  
engine = _detect(flowplayer.engines, (e) -> e.name == 'hlsjs-light')
engine.extensions.push(errorHandling)
```

**Comment by nnarhinen** [comment]
Thank you for taking the time to list this!

We'll tackle this to the next release!

**Comment by nnarhinen** [comment]
I agree that some kind of built-in error recovery should be implemented!

**Comment by codener** [comment]
Dear @nnarhinen, could you give an estimate on when error handling will arrive?

**Comment by nnarhinen** [comment]
Sorry for the delay @codener .

I expect 7.2.5 land next week.

**Comment by codener** [comment]
That'd be awesome. Keep up the good work!

**Comment by nnarhinen** [comment]
This landed yesterday in the dev branch. (There was a duplicate issue #1345 )

**Comment by codener** [comment]
Dear @nnarhinen, could you give another estimate on the release date? Thanks :)

**Comment by nnarhinen** [comment]
This is now releases as 7.2.5. Sorry for the problems in communication.

**Comment by codener** [comment]
Ok. Thanks for the hint though.

**Comment by fdambrosio** [comment]
when will be available 7.2.5 commercial version ? thanks

**Comment by bbbo** [comment]
it's already there, just not linked in the site while we're still updating the docs. Just replace the version number in the url...
